flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/table/KinesisDynamicTableFactoryTest.java

```diff
@@ -0,0 +1,358 @@
+		setProperty("aws.region", "us-west-2");
+	}};
+
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
+
+	// --------------------------------------------------------------------------------------------
+	// Positive tests
+	// --------------------------------------------------------------------------------------------
+
+	@Test
+	public void testGoodTableSource() {
+		TableSchema sourceSchema = defaultSourceSchema().build();
+		Map<String, String> sourceOptions = defaultSourceOptions().build();
+
+		// Construct actual DynamicTableSource using FactoryUtil
+		KinesisDynamicSource actualSource = (KinesisDynamicSource)
+			FactoryUtil.createTableSource(
+				null,
+				ObjectIdentifier.of("default", "default", SOURCE_TABLE),
+				createSourceTable(sourceSchema, sourceOptions, Collections.emptyList()),
+				new Configuration(),
+				Thread.currentThread().getContextClassLoader(),
+				false);
+
+		// Construct expected DynamicTableSink using factory under test
+		KinesisDynamicSource expectedSource = new KinesisDynamicSource(
+			sourceSchema.toPhysicalRowDataType(),
+			SOURCE_STREAM,
+			KINESIS_PROPERTIES,
+			new TestFormatFactory.DecodingFormatMock(",", true));
+
+		// verify that the constructed DynamicTableSink is as expected
+		assertEquals(expectedSource, actualSource);
+
+		// verify that the copy of the constructed DynamicTableSink is as expected
+		assertEquals(expectedSource, actualSource.copy());
+
+		// verify produced sink
+		ScanTableSource.ScanRuntimeProvider functionProvider =
+			actualSource.getScanRuntimeProvider(ScanRuntimeProviderContext.INSTANCE);
+		SourceFunction<RowData> sourceFunction =
+			as(functionProvider, SourceFunctionProvider.class).createSourceFunction();
+		assertThat(sourceFunction, instanceOf(FlinkKinesisConsumer.class));
+	}
+
+	@Test
+	public void testGoodTableSourceWithMetadataFields() {
+		TableSchema sourceSchema = defaultSourceSchema().build();
+		Map<String, String> sourceOptions = defaultSourceOptions().build();
+
+		// Construct actual DynamicTableSource using FactoryUtil
+		KinesisDynamicSource actualSource = (KinesisDynamicSource)
+			FactoryUtil.createTableSource(
+				null,
+				ObjectIdentifier.of("default", "default", SOURCE_TABLE),
+				createSourceTable(sourceSchema, sourceOptions, Collections.emptyList()),
+				new Configuration(),
+				Thread.currentThread().getContextClassLoader(),
+				false);
+
+		// Construct expected DynamicTableSink using factory under test
+		KinesisDynamicSource expectedSource = new KinesisDynamicSource(
+			sourceSchema.toPhysicalRowDataType(),
+			SOURCE_STREAM,
+			KINESIS_PROPERTIES,
+			new TestFormatFactory.DecodingFormatMock(",", true));
+
+		List<String> metadataKeys = Arrays.asList("ShardId", "ApproximateArrivalTimestamp");
+		DataType producedDataType = getProducedType(sourceSchema, Metadata.values());
+
+		expectedSource.applyReadableMetadata(metadataKeys, producedDataType);
```

Comment:
we are not really testing something if both call the same methods with the same parameters, instead we should test the contents of mutable member variables

Reply:
These are tested by the following `assertEquals(expectedSource, actualSource);` call, where `KinesisDynamicSource::equals` is

```
@Override
public boolean equals(Object o) {
	if (this == o) {
		return true;
	}
	if (o == null || getClass() != o.getClass()) {
		return false;
	}
	KinesisDynamicSource that = (KinesisDynamicSource) o;
	return Objects.equals(producedDataType, that.producedDataType) &&
		Objects.equals(requestedMetadataFields, that.requestedMetadataFields) &&
		Objects.equals(physicalDataType, that.physicalDataType) &&
		Objects.equals(stream, that.stream) &&
		Objects.equals(consumerProperties, that.consumerProperties) &&
		Objects.equals(decodingFormat, that.decodingFormat);
}
```

This goes somewhat [against the coding guideline](https://flink.apache.org/contributing/code-style-and-quality-java.html#equals--hashcode), but it is consistent with all other connectors (including `KafkaDynamicSource`). For me this is indicative of a broader issue with the ability interfaces. IMHO calls of ability interface methods should return modified copies of the original source/sink node as opposed to mutating that node (this is compatible with the current design of these classes as immutable factories).

Reply:
We had this discussions before. Personally, I was also more in favor of immutable interfaces. But after lots of online and offline discussion we decided for mutability by considering `DynamicTableSource` and `DynamicTableSink` as another level of factories or builders that the planner uses to get the runtime implementation in the end. 

Regarding this test, the `KinesisDynamicSource::equals` is correct but we should rather test the contents of the member variables instead of comparing the result of performing the same computation twice.

Btw I wouldn't say that the current implementation goes against the coding guidelines. The mutable attributes must be included in the hashCode/equals and it might be the case that the planner is using them.

Reply:
> Regarding this test, the KinesisDynamicSource::equals is correct but we should rather test the contents of the member variables instead of comparing the result of performing the same computation twice.

I see what you mean now, I will fix that.

> Btw I wouldn't say that the current implementation goes against the coding guidelines

I was referring to this part of the guidelines  

> When the methods take mutable fields into account, you often have a design issue.

and the possible design issue with the mutable factories.